src/core/SampleBuffer.cpp

```diff
@@ -131,6 +131,79 @@ SampleBuffer::SampleBuffer( const f_cnt_t _frames )
 
 
 
+SampleBuffer::SampleBuffer(SampleBuffer& orig):
+	m_audioFile(orig.m_audioFile),
+	m_origData(MM_ALLOC(sampleFrame, orig.m_origFrames)),
+	m_origFrames(orig.m_origFrames),
+	m_data(MM_ALLOC(sampleFrame, orig.m_frames)),
+	m_frames(orig.m_frames),
+	m_startFrame(orig.m_startFrame),
+	m_endFrame(orig.m_endFrame),
+	m_loopStartFrame(orig.m_loopStartFrame),
+	m_loopEndFrame(orig.m_loopEndFrame),
+	m_amplification(orig.m_amplification),
+	m_reversed(orig.m_reversed),
+	m_frequency(orig.m_frequency),
+	m_sampleRate(orig.m_sampleRate)
+{
+	orig.m_varLock.lockForRead();
+	//Deep copy m_origData and m_data from original
+	memcpy(m_origData, orig.m_origData, m_origFrames* BYTES_PER_FRAME);
+	memcpy(m_data, orig.m_data, m_frames* BYTES_PER_FRAME);
+	orig.m_varLock.unlock();
+}
+
+
+
+
+void swap(SampleBuffer& first, SampleBuffer& second) noexcept
+{
+	using std::swap;
+
+	// Lock both buffers for writing
+	// TODO: No guaranteed ordering, risk of deadlock?
+	first.m_varLock.lockForWrite();
+	second.m_varLock.lockForWrite();
+
+	// QString: built in swap method
+	first.m_audioFile.swap(second.m_audioFile);
+	// sampleFrame*: std::swap
+	swap(first.m_origData, second.m_origData);
+	swap(first.m_data, second.m_data);
+	// f_cnt_t: std::swap
+	swap(first.m_origFrames, second.m_origFrames);
+	swap(first.m_frames, second.m_frames);
+	swap(first.m_startFrame, second.m_startFrame);
+	swap(first.m_endFrame, second.m_endFrame);
+	swap(first.m_loopStartFrame, second.m_loopStartFrame);
+	swap(first.m_loopEndFrame, second.m_loopEndFrame);
+	// float: std::swap
+	swap(first.m_amplification, second.m_amplification);
+	swap(first.m_frequency, second.m_frequency);
+	// bool: std::swap
+	swap(first.m_reversed, second.m_reversed);
+	// sample_rate_t: std::swap
+	swap(first.m_sampleRate, second.m_sampleRate);
+
+	// Unlock again
+	first.m_varLock.unlock();
+	second.m_varLock.unlock();
+}
+
+
+
+
+SampleBuffer& SampleBuffer::operator=(SampleBuffer& that)
+{
+	SampleBuffer* temp = new SampleBuffer(that);
+	swap(*this, *temp);
```

Comment:
This allocates a `SampleBuffer` on the heap and leaks it. It would be better to allocate it on the stack.
```suggestion
	auto temp = that;
	swap(*this, temp);
```

Reply:
Isn't `auto temp = that` recursive, since this occurs inside the definition of assignment?

Reply:
`auto temp = that;` is an initialisation, not an assignment. Assignments put a new value into an existing object, but here a new object is being created.

Now I think about it though, a better option might be to take `that` by value. Then the copy is handled by the caller, and the existing copy constructor can take advantage of move semantics automatically as soon as a move constructor is defined.
```cpp
SampleBuffer& SampleBuffer::operator=(SampleBuffer that)
{
	swap(*this, that);
	return *this;
}
```